### PR TITLE
VM_bootloader:provide bootloader's version for each arch 

### PR DESCRIPTION
**Pull request by MiriamDeng**
ID:1456

### CODE REVIEW

**Comment by MiriamDeng** [comment]
Test results
 (2/2) Host_RHEL.m8.u9.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.9.0.ppc64le.io-github-autotest-qemu.boot: STARTED
 (2/2) Host_RHEL.m8.u9.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.9.0.ppc64le.io-github-autotest-qemu.boot: PASS (107.16 s)

**Comment by MiriamDeng** [comment]
@PaulYuuu @maxujun can you help to review this patch ? Thank you !

File: virttest/env_process.py [hunk 1]
```diff
@@ -1227,6 +1227,17 @@ def preprocess(test, params, env):
     LOG.debug("KVM userspace version(qemu): %s", kvm_userspace_version)
     version_info["qemu_version"] = str(kvm_userspace_version)
 
+    #Get the SLOF version for power pc
```
Comment (PaulYuuu) [inline virttest/env_process.py:1230]: For what I want is to let it be configurable, I have 2 ideas:
1. Get the firmware version for each platform:
    - ovmf, aarch64: check edk2 version
    - PC machine: check seabios
    - PPC: check SLOF
    - s390x: I don't know which one
2. Make all collection versions can be customizable, qemu/kvm/libvirt/swtpm/firmware...

Cc @luckyh @YongxueHong

File: virttest/env_process.py [hunk 2]
```diff
@@ -1227,6 +1227,17 @@ def preprocess(test, params, env):
...
+    slof_ver_cmd = params.get("slof_ver_cmd", "rpm -qa|grep SLOF")
+    if arch.ARCH in ('ppc64', 'ppc64le'):
+        if slof_ver_cmd:
+            try:
+                slof_version = a_process.run(
+                    slof_ver_cmd, shell=True).stdout_text.strip()
+                LOG.debug("SLOF version: %s", slof_version)
```
Comment (PaulYuuu) [inline virttest/env_process.py:1237]: Save to `version_info["firmware_version"]`

Comment (MiriamDeng) [reply virttest/env_process.py]: Updated, thanks.

**Comment by maxujun** [comment]
LGTM for power side.

File: virttest/env_process.py (hunk 1 above)
Comment (MiriamDeng) [reply virttest/env_process.py]: Hi all, 
I listed the build info for all arches and feel free to let me know if there's any missing.
1.
aarch64:  edk2
x86_64: both seabios and edk2
s390x: need to confirm with bfu
2.
swtpm for all arches.

**Comment by PaulYuuu** [comment]
After a short offline discussion with @YongxueHong, we think can only collect the boot firmware for each platform on this PR. A possible name may be `qemu_bios_version`? As SLOF/edk2/seabios is used to boot up a VM.

In short, in py file, log and save qemu_bios_version, in cfg:
```
default_bios:
    x86, i386:
        qemu_bios_ver_cmd = 
	aarch64:
        qemu_bios_ver_cmd = 
	ppc64le, ppc64:
        qemu_bios_ver_cmd = 
	s390x, s390:
        qemu_bios_ver_cmd = 
ovmf:
	qemu_bios_ver_cmd = 
```

What do you think @luckyh @MiriamDeng ?

**Comment by MiriamDeng** [comment]
Hi @PaulYuuu 
Thanks for the the guideline , will update the patch very soon. 
Thank you !

File: virttest/shared/cfg/base.cfg [hunk 3]
```diff
@@ -1087,3 +1087,10 @@ uuid_dimm = ""
 
 # Set image_extent_size_hint, displayed in qemu-img info extend_size_hint field.
 #image_extent_size_hint = 1G #To set extent_size_hint to 1G
+
+#
+#VM bootloader params
+#
+#Notes: please make sure the corresponding firmware before you set this value!
+#eg. for x86_64
+#bootloader = "rpm -qa|grep ^sea"
```
Comment (PaulYuuu) [inline virttest/shared/cfg/base.cfg:1096]: `vm_bootloader_ver_cmd = xxx`

**Comment by MiriamDeng** [comment]
Test result  (without internal cfg)
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.4.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (810.21 s

File: virttest/env_process.py [hunk 4]
```diff
@@ -1241,6 +1241,17 @@ def preprocess(test, params, env):
...
+    # Get the version of bootloader
+    vm_bootloader_cmd = params.get("vm_bootloader", "")
+    if vm_bootloader_cmd:
+        try:
+            vm_bootloader_ver = a_process.run(
+                vm_bootloader_cmd, shell=True).stdout_text.strip()
+            LOG.debug("vm_bootloader_version: %s", vm_bootloader_ver)
+            version_info["vm_bootloader_ver"] = str(vm_bootloader_ver)
+        except a_process.CmdError:
+            vm_bootloader_ver = "Unkown"
```
Comment (PaulYuuu) [inline virttest/env_process.py:1253]: Even if the version is "Unknown", it can also be saved in the version_info, like qemu/libvirt doing.

Comment (MiriamDeng) [reply virttest/env_process.py:1251]: Thanks, updated

**Comment by MiriamDeng** [comment]
Also created internal cfg #1241

File: virttest/env_process.py [hunk 5]
```diff
@@ -1241,6 +1241,17 @@ def preprocess(test, params, env):
             test.cancel("Got host qemu version:%s, which is not in %s" %
                         (host_qemu, required_qemu))
 
+    # Get the version of bootloader
+    vm_bootloader_cmd = params.get("vm_bootloader", "")
```
Comment (YongxueHong) [inline virttest/env_process.py:1245]:
```suggestion
    vm_bootloader_cmd = params.get("vm_bootloader_ver_cmd", "")
```

File: virttest/env_process.py [hunk 6]
```diff
@@ -1241,6 +1241,17 @@ def preprocess(test, params, env):
...
+    if vm_bootloader_cmd:
+        try:
+            vm_bootloader_ver = a_process.run(
+                vm_bootloader_cmd, shell=True).stdout_text.strip()
+        except a_process.CmdError:
+            vm_bootloader_ver = "Unkown"
+        version_info["vm_bootloader_ver"] = str(vm_bootloader_ver)
+        LOG.debug("vm_bootloader_version: %s", vm_bootloader_ver)
```
Comment (YongxueHong) [inline virttest/env_process.py:1253]:
```suggestion
        LOG.debug("VM bootloader version: %s", vm_bootloader_ver)
```

Comment (MiriamDeng) [reply virttest/env_process.py]: Updated,thanks !

File: virttest/env_process.py [hunk 7]
```diff
@@ -1244,6 +1244,17 @@ def preprocess(test, params, env):
...
 
+    # Get the version of bootloader
+    vm_bootloader_cmd = params.get("vm_bootloader_ver_cmd", "")
+    if vm_bootloader_cmd:
+        try:
+            vm_bootloader_ver = a_process.run(
+                vm_bootloader_cmd, shell=True).stdout_text.strip()
+        except a_process.CmdError:
+            vm_bootloader_ver = "Unkown"
+        version_info["vm_bootloader_ver"] = str(vm_bootloader_ver)
```
[the remark this answers is not in the document]
Comment (MiriamDeng) [reply virttest/env_process.py]: Updated, thanks

**Comment by MiriamDeng** [comment]
Test results with internal cfg
 (1/3) Host_RHEL.m8.u9.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.xxx.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
 (1/3) Host_RHEL.m8.u9.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.xxx.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (967.83 s)

File: virttest/env_process.py [hunk 8]
```diff
@@ -1244,6 +1244,17 @@ def preprocess(test, params, env):
             test.cancel("Got host qemu version:%s, which is not in %s" %
                         (host_qemu, required_qemu))
 
+    # Get the version of bootloader
+    vm_bootloader_cmd = params.get("vm_bootloader_ver_cmd", "")
```
Comment (YongxueHong) [inline virttest/env_process.py:1248]:
```suggestion
    vm_bootloader_ver_cmd = params.get("vm_bootloader_ver_cmd", "")
```

**Review by YongxueHong — APPROVED** [review]

**Review by PaulYuuu — APPROVED** [review]

**Pull request merged**